File: LinkedList/SinglyLinkedList.cpp
#include <iostream>

class Node{
    public:
        //Constructors
        Node() : val(0), next(nullptr) {}
        Node(double init_val, Node* init_next = nullptr): val(init_val), next(init_next) {}
        //Deconstructor
        ~Node(){}
        //Methods
        double getValue(){return val;}
        void setValue(double val){this->val = val;}
        Node* getNext(){return next;}
        void setNext(Node* Nextnode){next = Nextnode;}
    private:
        double val;
        Node* next;
};

class SinglyLinkedList{
    private:
        Node* head;
        Node* tail;
        int length;
    public:
        //Constructor
        SinglyLinkedList(Node* init_head=nullptr, Node* init_tail=nullptr, int init_length=0): head(init_head), tail(init_tail), length(init_length) {}
        //Deconstructor (Free the memory)
        ~SinglyLinkedList(){while (head != nullptr) {remove_head();}}
        
        void head_insert(double data){
            Node* new_node = new Node(data);
            new_node->setNext(head);
            head = new_node;
            length += 1;
            if(tail == nullptr){tail=new_node;}
        }
    
        void tail_insert(double data){
            Node* new_node = new Node(data);
            if(tail == nullptr){
                head = tail = new_node;
            }
            else{
                tail->setNext(new_node);
                tail = new_node;
            }
            length += 1;
        }
    
        // Insert before N-th node
        void insert(int N, double data){
            if (N < 1){
                std::cout<<"Error! Position of insertion N="<<N<<" has to be a postive integer!\n";
            }
            else if (N > length + 1){
                std::cout<<"Error! Position of insertion N="<<N<<" exceeds the length (L="<<length<<") of current SinglyLinkedList!\n";
            }
            else if (N == 1){
                head_insert(data);
            }
            else if (N == length + 1){
                tail_insert(data);
            }
            else{
                Node* previous = head;
                Node* current = previous->getNext();
                Node* new_node = new Node(data);
                while(N>2){
                    previous = current;
                    current = current->getNext();
                    N -= 1;
                }
                previous->setNext(new_node);
                new_node->setNext(current);
                length += 1;
            }
        }
    
        void remove_head(){
            if(head){
                Node* oldHead = head;
                head = head->getNext();
                // Free the memory
                delete  oldHead;
                length -= 1;
            }
            //After excute above step, if SLL becomes None, tail also needs to be set to None
            if (head == nullptr){tail = nullptr;}
        }
    
        //remove N-th node
        void remove(int N){
            if(head == nullptr){return;}
            if(N > length){std::cout<<"Error! Try to delete " << N <<"-th node, but it does not exist!\n";return;}
            if(N == 1){remove_head();}
            else {
                Node* previous = head;
                Node* current = previous->getNext();              
                while(N > 2){
                    previous = current;
                    current = current->getNext();
                    N -= 1;
                }
                previous->setNext(current->getNext());
                // If the tail node is deleted, we need to update the tail
                if(current->getNext() == nullptr){tail=previous;}
                length -= 1;
                // Free the memory
                delete current;
            }
            
        }
     
        void print_list(){
            std::cout<<"Length:"<<length<<std::endl;
            Node* current = head;
            while(current){
                std::cout<<current->getValue()<<"→";
                current=current->getNext();
            }
            std::cout<<"\n";
        }
    
};

int main(){
    SinglyLinkedList sll;
    sll.head_insert(2);
    sll.head_insert(3);
    sll.head_insert(4);
    sll.head_insert(5);
    sll.print_list();
    sll.tail_insert(1);
    sll.print_list();
    sll.insert(3,777);
    sll.print_list();
    sll.insert(10,777);
    sll.insert(-1,777);
    sll.remove_head();
    sll.print_list();
    sll.remove(2);
    sll.print_list();
    sll.remove(4);
    sll.print_list();
    sll.remove(4);
    sll.remove(1);
    sll.remove(2);
    sll.print_list();
    sll.remove(1);
    sll.print_list();
    return 0;
}

File: LinkedList/SinglyLinkedList.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None
        
class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
        
    def head_insert(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
        if self.tail is None:
            self.tail = new_node
        self.length += 1
        
    def tail_insert(self, data):
        new_node = Node(data)
        if self.tail is None:
            self.head = self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.length += 1
        
    def insert(self, N, data):
        if N <1 or not isinstance(N,int):
            raise ValueError("Error! Position of insertion N="+str(N)+" has to be a postive integer!\n")
        if N > self.length+1:
            raise ValueError("Error! Position of insertion N="+str(N)+" exceeds the length (L="+str(self.length)+") of current SinglyLinkedList!\n")
        if N == 1:
            self.head_insert(data)
            return
        if N == self.length+1:
            self.tail_insert(data)
            return
        previous = self.head
        current = previous.next
        new_node = Node(data)
        while N > 2:
            previous, current = current, current.next
            N -= 1
        previous.next = new_node
        new_node.next = current
        self.length += 1        
        
    def remove_head(self):
        if self.head is not None:
            # self.head is automatically reclaimed by the garbage collector if no longer referenced
            self.head=self.head.next
            self.length -= 1
        # After excute above step, if SLL becomes None, self.tail also needs to be set to None
        if self.head is None:
            self.tail = None
        
    def remove(self, N):#remove N-th node
        if self.head is None:
            return
        if N > self.length:
            raise ValueError("Error! Try to delete " + str(N) + "-th node, but it does not exist!\n")
        if N ==1:
            self.remove_head()
        else:
            previous = self.head
            current = previous.next
            while N > 2:
                previous, current = current, current.next
                N -= 1
            previous.next = current.next
            # If the tail node is deleted, we need to update the tail
            if current.next is None:
                self.tail = previous
            self.length -= 1
          
    def print_list(self):
        print("Length:",self.length)
        current=self.head
        while current:
            print(current.data, end="→")
            current=current.next
        print("\n")
        
        
#Test            
sll = SinglyLinkedList()

sll.head_insert(3)
sll.head_insert(2)
sll.head_insert(1)
print("After head inserts:")
sll.print_list()

sll.tail_insert(4)
sll.tail_insert(5)
print("After tail inserts:")
sll.print_list()

sll.insert(3, 2.5)
print("After inserting 2.5 at position 3:")
sll.print_list()

sll.remove_head()
print("After removing head:")
sll.print_list()

sll.remove(3)
print("After removing the element at position 3:")
sll.print_list()

try:
    sll.insert(-1, 0)
except ValueError as e:
    print(e)

try:
    sll.remove(10)
except ValueError as e:
    print(e)
    
try:
    sll.insert(3.5, 0)
except ValueError as e:
    print(e)
